### ISSUE TITLE
A Way to Unsubscribe Agents from an Address?

### ISSUE THREAD

**Issue by sjanko2** [opensistemas-hub/osbrain#175]
Hi again,

I'm hoping to create temporary groups of agents during each time step of my simulation, and was planning on using a temporary publish/subscribe communication address to form this group. However, I can't seem to figure out a way to unsubscribe or remove an agent from the communication address once it has been added. 

I have been reading through the osBrain Library API and haven't found anything about it yet. Is there a way to do this? If not, is there a better way to group agents together?

Thanks so much for your help, as always!

Sam

**Comment by Peque** [comment]
@sjanko2 If you are trying to unsubscribe, we are about to change and improve the API for that (right now it is a bit limited). Anyway, if that is the case, I could tell you how to do it.

If you just want to close a connection, then try: `agent.close('alias')`

Does that work for you? If it does not, could you provide a reduced code sample?

Can we close #167?

**Comment by Peque** [comment]
Opened a related issue: #176

Thanks for pointing us missing documentation again. :blush:

**Comment by sjanko2** [comment]
Hi there! 

Thank you very much for your response. Yes, feel free to close #167 , my apologies for not clarifying that sooner. The solution worked! :)

For this issue, I have attempted your solution in the attached example code. In this code, two "groups" are formed with different communication addresses. The main group allows Alice to send messages to agents Bob, Eve, and Dave. The temporary group allows Alice to send messages to agents Bob and Eve only.

After this is set up, two messages are sent to the main group followed by two messages sent to the temporary group. I then attempt to close the connection with your suggested solution: `agent.close('alias')` . This produces an error, which I have attached an image of.

What I would like to do is unsubscribe all agents from this "Tempaddr" so that when a message is sent from Alice the second time, no one receives it. However, the messages sent from Alice to the main group should continue to work.

Thanks so much!!

Sam

---

UnsubscribeExample.py

---

```
import time
import osbrain
from osbrain import run_agent
from osbrain import run_nameserver

osbrain.config['TRANSPORT'] = 'tcp'

def log_Everyone(agent, message):
    agent.log_info('The message for everyone is: %s' % message)

def log_TemporaryGroup(agent, message):
    agent.log_info('The temporary group message is: %s' % message)

if __name__ == '__main__':

    # Start the nameserver:
    run_nameserver('127.0.0.1:1124')

    alice = run_agent('Alice')
    bob = run_agent('Bob')
    eve = run_agent('Eve')
    dave = run_agent('Dave')

    # System configuration
    addr = alice.bind('PUB', alias='main')
    bob.connect(addr, handler={'MainGroup': log_Everyone})
    eve.connect(addr, handler={'MainGroup': log_Everyone})
    dave.connect(addr, handler={'MainGroup': log_Everyone})

    Tempaddr = alice.bind('PUB', alias='TempAddr')
    bob.connect(Tempaddr, handler={'TemporaryGroup': log_TemporaryGroup})
    eve.connect(Tempaddr, handler={'TemporaryGroup': log_TemporaryGroup})

    # Send messages to all three other agents
    for _ in range(2):
        time.sleep(1)
        message = 'Hello, Everyone!'
        alice.send('main', message, topic='MainGroup')

    # Send messages to only those a part of the temporary group
    for _ in range(2):
        time.sleep(1)
        message = 'This is a message to the temporary group only.'
        alice.send('TempAddr', message, topic='TemporaryGroup')

    time.sleep(1)
    alice.close('TempAddr')

    # Test to see if this successfully closed the group:
    # Send messages to all three other agents (this should still work)
    for _ in range(2):
        time.sleep(1)
        message = 'Hello, Everyone!'
        alice.send('main', message, topic='MainGroup')

    # Send messages to only those a part of the temporary group (this should error)
    for _ in range(2):
        time.sleep(1)
        message = 'This is a message to the temporary group only.'
        alice.send('TempAddr', message, topic='TemporaryGroup')
```
![unsubscribeerror](https://user-images.githubusercontent.com/9323425/28554811-7f06f656-70b0-11e7-8b46-be861af6bbeb.PNG)

**Comment by Flood1993** [comment]
Related:
#158 if merged, will give the user more control over subscriptions/handlers.
#177 if merged, will give the user more control on closing connections (sockets).

Hopefully they will be reviewed by tomorrow.

**Comment by Peque** [comment]
Merged #177. #158 will probably be merged soon (maybe this week).

@sjanko2 you might be interested in trying the latest osBrain code available in the master branch to be able to [close your connections easily](https://osbrain.readthedocs.io/en/latest/considerations.html#closing-connections). Otherwise, we will probably release a new osBrain version by the end of next week.

Thanks for your feedback!

**Comment by Peque** [comment]
Closing this as #158 was already merged.
